### ISSUE TITLE
Serialplot won't refresh

### ISSUE THREAD

**Issue by schwukas** [hyOzd/serialplot#5]
Hi there, weird issue:

I have serialplot running on my T430 without any problems. Plugging the same board into my Desktop PC, I only can get new data whenever I press "Open". So it appears that serialplot isn't polling for new data and I have to "refresh" it manually. My T430 is running Arch Linux, my Desktop is on Manjaro. Both have qwt 6.1 and the complete qt5 package installed. 

Is this a known problem and is there a fix? I'm also happy to provide additional information, bug reports whatever :). I didn't have time to dig into the code, but I'll gladly free up some, if it helps.

**Comment by hyOzd** [comment]
@schwukas I've never heard of this one before. This sounds like an issue with the Qt library or maybe even deeper issue (driver, kernel, hw etc).

Can you tell me what kind of serialport are you using on your PCs. USB-Uart converter etc.? If you are using motherboard RS232 connector and your desktop maybe you can try a USB converter?

Also which build of SerialPlot are you using. Did you build it yourself? What is your Qt version? Have you tried running AppImage version?

**Comment by schwukas** [comment]
What a response time :astonished: 

I'm using a cheap generic USB to TTL dongle on both systems.

* `serialplot -v` returns `0.11.0`. The AUR package I used to install it seems to build commit `834.866d3c2d410a-1`. It's just cloning the latest code from the sourcehut repo. 
* I did not build it myself but used the "ready-made" version from the [Arch User Repository](https://aur.archlinux.org/packages/serialplot-hg/). I can build it myself later, if that helps.
* I'm on `Qt 5.15.1 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 10.2.0) on "xcb"` on both systems.
* I have tried the AppImage version taken from [hackaday.io](https://hackaday.io/project/5334-serialplot-realtime-plotting-software). Same issue there.
* I also tried everything once again with root, but that didn't make a difference.

**Comment by hyOzd** [comment]
Have you tried another software, a serial port terminal? 

What data format are you using? Have you tried another format, just select another format from "Data Format" tab, without changing your input. You should potentially see at least gibberish data plotted (if there is a problem in SerialPlots parsers).

**Comment by schwukas** [comment]
Yes, sorry I haven't mentioned that it's working fine in both the Arduino Serial Plotter and cutecom.

I first have copied all the values from cutecom to serialplot to make sure it's the same. When that didn't work I just played around with every possible combination without a result. When I click on "open" and receive that one little bit of data, everything looks fine. The data makes sense, the plot is drawn correctly. So I'd _guess_ that my settings are ok.

![screenshot-20201102-113849](https://user-images.githubusercontent.com/33559975/97858808-fd7d0800-1cff-11eb-8300-96eb66373409.png)
![screenshot-20201102-113902](https://user-images.githubusercontent.com/33559975/97858832-07067000-1d00-11eb-9cab-dbc2e7ab6bee.png)

**Comment by hyOzd** [comment]
> Forget to mention as well: I can get the plot to draw something when using the "Simple Binary" but the values are just jumping around randomly.

This is actually good news. It means there is probably a problem with SerialPlot ascii parser. Which means it is *probably* something we can solve : )

Do you see any errors on the "Logs" tab?

And can you post a piece of your data that is captured with another tool? Maybe you can also let me know your code that prints the data (if it is simple)?

**Comment by schwukas** [comment]
No errors in the debug log. Just the opening and closing messages.

I made a simple saw tooth that's hopefully easier to analyse.
```
void setup() {
  Serial.begin(115200);
}

void loop() {
  for (int i = 0; i < 250; i++) {
    Serial.println(i);
    delay(5);
  }
}
```

So values from 0 to 250 in increments of 1 are expected.

I'm sure you know what a sawtooth looks like but for completeness' sake:
![screenshot-20201102-125307](https://user-images.githubusercontent.com/33559975/97865385-65d0e700-1d0a-11eb-8845-98737199be63.png)

**Comment by hyOzd** [comment]
This is a difficult problem for me to reproduce. I need your help.. I've prepared a custom debug image that prints some messages. Can you run this and let me know its command line output while problem is happening? Maybe switch between data formats and see what happens? I suggest you run it with a small number of samples per seconds as it will print a lot of lines.

https://nextcloud.ozderya.net/index.php/s/n9reAk7WsjAtobr

**Comment by schwukas** [comment]
ASCII - Column Delimiter: comma

```
[Debug] SerialPlot 0.11.0
[Debug] Revision 866d3c2d410a+
[Debug] Opened port: "ttyACM0"
[Debug] on readData
[Debug] can read line
[Debug] bytes: "14\r\n"
[Debug] firstReadAfterEnable
[Debug] can read line
[Debug] bytes: "41\r\n"
[Debug] numComingChannels:  1
[Debug] samples nc:  1  size: 1
[Debug] numBytesRead: 8
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] Closed port: "ttyACM0"
```
ASCII - Column delimiter: space

```
[Debug] SerialPlot 0.11.0
[Debug] Revision 866d3c2d410a+
[Debug] Opened port: "ttyACM0"
[Debug] on readData
[Debug] can read line
[Debug] bytes: "4\r\n"
[Debug] firstReadAfterEnable
[Debug] can read line
[Debug] bytes: "5\r\n"
[Debug] numComingChannels:  1
[Debug] samples nc:  1  size: 1
[Debug] numBytesRead: 6
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] on readData
[Debug] numBytesRead: 0
[Debug] Closed port: "ttyACM0"
```

Using the Binary Format just shows opening and closing the port, nothing else.
If there's anything else I can help you with, just let me know.

**Comment by hyOzd** [comment]
Well this is going to be tricky. Unfortunately this looks like a limitation/bug of Qt. Every time new data come from device this function is called:

```
unsigned AsciiReader::readData()
{
    unsigned numBytesRead = 0;

    while(_device->canReadLine()) {
        ... read data here ...
    }
```

Problem looks like readData is called correctly, but `canReadLine` (a Qt function) doesn't return true. I suspect this is a problem with buffering. Why it doesn't happen on your other device but happens on this one I have no idea. I suspect driver versions are different.

Can you please try these 2 build as well. 2nd one only adds a couple debug messages. Can you post its output just like before?

2nd: https://nextcloud.ozderya.net/index.php/s/zFcDP7yZzS6Atts

This 3rd build does something different. It clears buffers immediately when port is opened. If it is a buffering problem as I suspect, this might fix the issue. You don't have to post its output if it behaves the same. 

3rd: https://nextcloud.ozderya.net/index.php/s/WzXLNPqpwPx9pG7

**Comment by hyOzd** [comment]
@schwukas well this is weird. There is something else in the input data. 

`0x80`, `0x00` these are not valid ASCII characters. They are usually used for control stuff. I also noticed `0xFE` which is usually used as "Frame Start Byte" in packet data formats. Are you sure your device isn't sending something else to serial port? Or maybe you have a problem in your connection and we are seeing junk?

Can you please check again with a serialport terminal in raw (hex) mode and see if you can spot similar data as above?

**Comment by schwukas** [comment]
It took a while for me to get around to it but I checked the output with cutecom:

```
00000000 30 0d 0a 30 0d 0a 31 0d   0a 32 0d 0a 33 0d 0a 34 	0␍␊0␍␊1␍  ␊2␍␊3␍␊4
00000016 0d 0a 35 0d 0a 36 0d 0a   37 0d 0a 38 0d 0a 39 0d 	␍␊5␍␊6␍␊  7␍␊8␍␊9␍
00000032 0a 31 30 0d 0a 31 31 0d   0a 31 32 0d 0a 31 33 0d 	␊10␍␊11␍  ␊12␍␊13␍
00000048 0a 31 34 0d 0a 31 35 0d   0a 31 36 0d 0a 31 37 0d 	␊14␍␊15␍  ␊16␍␊17␍
00000064 0a 31 38 0d 0a 31 39 0d   0a 32 30 0d 0a 32 31 0d 	␊18␍␊19␍  ␊20␍␊21␍
```

This looks like valid ASCII to me. I then turned off cutecom and immediately recorded in serialplot but received the same junk from above. Weird..

**Comment by hyOzd** [comment]
@schwukas This is my bad. Debug executable that I provided to you has a piece of code that read the data from the port that shouldn't be read assuming it was junk data. That is why parser isn't working.

But input data seems to be correct in this case. Are you still having problem with release executable?

**Comment by schwukas** [comment]
No worries :relaxed: 

> Are you still having problem with release executable?
Unfortunately, none of the three AppImages I have tried, fixed my issue. I'm okay to accept it's a hardware thing. I chose a crappy, cheap motherboard when I built this PC and regretted it on several occasions. Maybe that's just another one of those occasions. Although I would love to use Serialplot, as it offers way more flexibility than Arduino and Cutecom. Oh well.. :slightly_smiling_face:
